MagazineLayout/Public/MagazineLayout.swift

```diff
@@ -721,29 +724,60 @@ public final class MagazineLayout: UICollectionViewLayout {
-
-    if isScrolling && isSizingElementAboveTopEdge {
+    let top = currentCollectionView.bounds.minY + contentInset.top
+    let bottom = currentCollectionView.bounds.maxY
+    let isSizingElementAboveTopEdge = originalAttributes.frame.minY < top
+    let isSizingElementBelowBottomEdge = originalAttributes.frame.minY > bottom
+
+    let contentOffsetAdjustment: CGPoint
+    if verticalLayoutDirection == .topToBottom, isSizingElementAboveTopEdge {
+      // If layout information is discarded above our current scroll position (on rotation, for
+      // example), we need to compensate for preferred size changes to items as we're scrolling up,
+      // otherwise, the collection view will appear to jump each time an element is sized.
+      // Since size adjustments can occur for multiple items in the same soon-to-be-visible row, we
+      // need to account for this by considering the preferred height for previously sized elements
+      // in the same row so that we only adjust the content offset by the exact amount needed to
+      // create smooth scrolling.
       let isSameRowAsLastSizedElement = lastSizedElementMinY?.isEqual(
         to: currentElementY,
         threshold: 1 / scale)
         ?? false
+
       if isSameRowAsLastSizedElement {
         let lastSizedElementPreferredHeight = self.lastSizedElementPreferredHeight ?? 0
         if preferredAttributes.size.height > lastSizedElementPreferredHeight {
-          context.contentOffsetAdjustment.y = preferredAttributes.size.height - lastSizedElementPreferredHeight
+          contentOffsetAdjustment = CGPoint(
+            x: 0,
+            y: preferredAttributes.size.height - lastSizedElementPreferredHeight)
+        } else {
+          contentOffsetAdjustment = .zero
         }
       } else {
-        context.contentOffsetAdjustment.y = preferredAttributes.size.height - originalAttributes.size.height
+        contentOffsetAdjustment = CGPoint(
+          x: 0,
+          y: preferredAttributes.size.height - originalAttributes.size.height)
       }
+    } else if
+      verticalLayoutDirection == .bottomToTop,
+      !isSizingElementBelowBottomEdge,
+      modelState.isPerformingBatchUpdates
+    {
+      contentOffsetAdjustment = CGPoint(
+        x: 0,
+        y: preferredAttributes.size.height - originalAttributes.size.height)
+    } else {
+      contentOffsetAdjustment = .zero
+    }
+
+    if modelState.isPerformingBatchUpdates {
+      // If we're in the middle of a batch update, we need to adjust our content offset. Doing it
+      // here in the middle of a batch update gets ignored for some reason. Instead, we delay
+      // slightly and do it in `finalizeCollectionViewUpdates`.
+      // This needs to be set in `finalizeCollectionViewUpdates`, otherwise it doesn't get applied.
+      stagedContentOffsetAdjustment = contentOffsetAdjustment
+    } else if isScrolling {
+      context.contentOffsetAdjustment = contentOffsetAdjustment
```

Comment:
this code was already barely intelligible trash, and this certainly didn't help. It works though, and handles the cases where self-sizing items are inserted at the top / bottom